src/emulator/instruction.ts

```diff
@@ -27,32 +27,27 @@ export const enum Operation {
 }
 
 export const enum AddressingMode {
-    implicit,
-    imm16,
     imm8,
+    imm8ind,
+    imm8io,
     reg8,
-    reg16,
+    reg8io,
     ind8,
-    reg16_imm16,
-    reg8_imm8,
-    imm8_reg8,
-    immind8_reg8,
-    ind8_reg8,
-    ind8_imm8,
-    reg8_ind8,
-    reg8_reg8,
-    imm8io_reg8,
-    reg8_imm8io,
-    reg8io_reg8,
-    reg8_reg8io,
+
+    imm16,
+    reg16,
+}
+
+export interface OperationParameter {
+    value?: number;
+    addressingMode: AddressingMode;
 }
 
 export interface Instruction {
     opcode: number;
     operation: Operation;
-    addressingMode: AddressingMode;
-    par1: number;
-    par2: number;
+    par1?: OperationParameter;
```

Comment:
I would like to make `par1` and `par2` non-optional and just preset them to `implicit/0` in order to keep this data structure as static as possible (it is used in the hot code path).

Reply:
Another thing to think about: once `par1` and `par2` are not optional anymore the additional object could be removed and be reduced to plain fields `par1/2` and `mode1/2` again. This is less clean, but it would avoid an additional dereference on every instruction decode.